### ISSUE TITLE
Recolour SVG image

### ISSUE THREAD

**Issue by chenyu-psy** [kurokida/jspsych-psychophysics#45]
Hello,

I wonder if there is an easy way to change the color of an SVG image. In my experiment, I want to display several object with varied colour. The colour is assigned according to the experimental manipulation.

At the moment, I have tried to display the image with image stimulus or using manual stimulus to display it. When I use the `image` stimulus, I did not find an argument to change the color of the image. And when I tried to use `drawFunc`, I just found that the image did not display even if I didn't code anything in the `drawFunc`.

In this case, I started to use `manual` stimulus. Here is my code:

```javascript
const iconObject = {
        obj_type: 'manual',
        drawFunc: function (stimulus, canvas, context) {
            // Create a new image element
            var img = new Image();

            // Set the source of the image to the provided imgPath
            img.src = file;

            // When the image is loaded, draw it on the canvas
            img.onload = function () {
                // Clear the canvas
                context.clearRect(0, 0, canvas.width, canvas.height);
                // Draw the image on the canvas
                context.drawImage(img, 0, 0, canvas.width, canvas.height);
            }
        }
```
I found even if I only want to display the image, it did not work.

It would be very helpful if there were an easy solution to do it. Thank you in advance!

**Comment by chenyu-psy** [comment]
Thank you for your quick response!

In my experiment, I want to ask participants to remember the colour of a specific object. I will present the object and ask participants to select the correct colour. The picture below is an example.

![image](https://github.com/kurokida/jspsych-psychophysics/assets/57704488/91c31947-96ce-4f53-a946-61725bc65bf3)

This task requires me to change the color of the picture. However, I really don't know how to do it. I asked chatGPT, and it showed me several different ways:

If my SVG image is inline within an HTML document:
```html
<svg id="mySVG" width="100" height="100">
    <circle cx="50" cy="50" r="40" fill="blue" />
</svg>

```

Or using JavaScript and the DOM
```typescript
// Fetch the SVG file
fetch('your-svg-image.svg')
    .then(response => response.text())
    .then(svgText => {
        // Parse the SVG content
        const parser = new DOMParser();
        const svgDoc = parser.parseFromString(svgText, 'image/svg+xml');
        const svgElement = svgDoc.documentElement;

        // Change the fill color of a specific element in the SVG
        const circleElement = svgElement.querySelector('circle');
        circleElement.setAttribute('fill', 'green');

        // Append the modified SVG back to your document
        document.body.appendChild(svgElement);
    });

``` 

Or using the D3.js package:
``` typescript
// Assuming you have already included D3.js library in your HTML
const svg = d3.select('#mySVG'); // Select the SVG element

// Change the fill color of a specific element in the SVG
svg.select('circle')
   .attr('fill', 'purple');

```

However, I still don't know how to combine these solution with the jspsych plugins or the psychophysics plugin.

**Comment by kurokida** [comment]
Thank you for your explanation.

Actually, I had never used SVG files before, and my plugin does not work well with SVG files.

But, if you don't need to my plugin, the following code might be helpful.

```javascript
  var jsPsych = initJsPsych({
    on_finish: function() {
        jsPsych.data.displayData();
    }
  });

  var timeline = [];

  const trial = {
    timeline: [
        {
            type: jsPsychHtmlButtonResponse,
            stimulus: function(){
                // Open your SVG file in a text editor, and then copy and paste the <svg> element.
                return `<svg version="1.1" id="mysvg" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="500px" height="500px" x="0px"
                    y="0px" viewBox="0 0 500 500" style="enable-background:new 0 0 500 500;" xml:space="preserve">
                <style type="text/css">
                    .st0{fill: ${jsPsych.timelineVariable('rect_color')};}
                    .st1{fill: ${jsPsych.timelineVariable('circle_color')};}
                    .st2{fill: ${jsPsych.timelineVariable('star_color')};}
                </style>
                <g>
                    <rect x="75.5" y="96.5" class="st0" width="87" height="65"/>
                    <path d="M162,97v64H76V97H162 M163,96H75v66h88V96L163,96z"/>
                </g>
                <g>
                    <ellipse class="st1" cx="300.5" cy="340" rx="62" ry="59.5"/>
                    <path d="M300.5,281c33.91,0,61.5,26.47,61.5,59s-27.59,59-61.5,59S239,372.53,239,340S266.59,281,300.5,281 M300.5,280
                        c-34.52,0-62.5,26.86-62.5,60s27.98,60,62.5,60s62.5-26.86,62.5-60S335.02,280,300.5,280L300.5,280z"/>
                </g>
                <g>
                    <polygon class="st2" points="282.03,199.6 239.34,180.06 280.81,158.07 286.2,111.44 319.93,144.08 365.95,134.8 345.33,176.97 
                        368.38,217.86 321.9,211.28 290.13,245.84 	"/>
                    <path d="M286.58,112.5l32.82,31.75l0.38,0.37l0.52-0.1l44.76-9.03l-20.06,41.02l-0.23,0.47l0.26,0.46l22.42,39.78l-45.21-6.4
                        l-0.52-0.07l-0.36,0.39l-30.9,33.62l-7.88-44.98l-0.09-0.52l-0.48-0.22l-41.52-19l40.34-21.4l0.46-0.25l0.06-0.52L286.58,112.5
                        M285.82,110.38l-5.48,47.37l-42.13,22.35l43.36,19.85l8.23,46.97l32.28-35.11l47.22,6.68l-23.42-41.55l20.95-42.84l-46.75,9.43
                        L285.82,110.38L285.82,110.38z"/>
                </g>
                </svg>
                `
            },
            choices: ['Red', 'Green', 'Blue'],
        }
    ],
    timeline_variables: [ // There are two trials. You can specify the stimulus colors as you like.
        {rect_color: "red", circle_color: "blue", star_color: "yellow"},
        {rect_color: "#ff00ff", circle_color: "#98fb98", star_color: "#ffd700"}
    ],
  }

  timeline.push(trial)

  jsPsych.run(timeline);
```
Best regards,
Daiichiro

**Comment by chenyu-psy** [comment]
Thank you for your code!

I spent some time implementing the code in my experiment. Because there are more than 20 pictures that I want to display. Copying and pasting the SVG elements for each picture is not very convenient. 

Fortunately, I found an another way to do it. I will show my code here in case other people have the same issue.
1. Prepare some square pictures where the object is transparent but the background is a solid colour.
2. Every time you want to display a picture, display a square of the same size as the picture behind it.
3. Because the object is transparent, when you change the colour of the square, the transparent object will be filled with that colour.

```typescript
interface ImageObject {
    imagePath: string;
    color?: string;
    width: number;
    position: number[];
  }

export function drawImageObject(images:ImageObject[]) {

    let objectList:any[] = [];

    images.forEach((imageObj) => {
        const { imagePath, color, width, position  } = imageObj;

        var rect_object = {
            obj_type: 'rect',
            width: width,
            height: width,
            fill_color: color,
            line_color: color,
            startX: position[0],
            startY: position[1]
        }
    
        var img_object = {
            obj_type: 'image', 
            file: imagePath,
            scale: (width + 3)/SIZE.REAL_IMAGE_SIZE, // the real size of the image is a global variable
            startX: position[0],
            startY: position[1]
        }

        objectList.push(rect_object);
        objectList.push(img_object);
        
    });

    return objectList;
}
```

**Comment by kurokida** [comment]
Thank you very much for sharing your solusion.

I made two sample programs for SVG files. 

https://github.com/kurokida/jspsych-psychophysics/blob/svg_test/psychophysics-demos/svg_change_color1.html
This is a code which implemented your idea although I deleted proprietary code in TypeScript.

https://github.com/kurokida/jspsych-psychophysics/blob/svg_test/psychophysics-demos/svg_change_color2.html
This is a code which implemented my solution. As you mentioned, my idea is not suitbale when a lot of SVG files are used. But, it would be useful to change the color of complex stimuli.
